test/ServiceTest.cpp

```diff
@@ -0,0 +1,107 @@
+//  Copyright 2022, University of Freiburg,
+//  Chair of Algorithms and Data Structures.
+//  Author: Hannah Bast <bast@cs.uni-freiburg.de>
+
+#include <gtest/gtest.h>
+
+#include "./HttpTestHelpers.h"
+#include "./IndexTestHelpers.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+#include "ctre/ctre.h"
+#include "engine/Service.h"
+#include "parser/GraphPatternOperation.h"
+#include "util/http/HttpUtils.h"
+
+// _____________________________________________________________________________
+TEST(ServiceTest, computeResult) {
+  // Query execution context (with small test index) and allocator for testing,
+  // see `IndexTestHelpers.h`.
+  using ad_utility::AllocatorWithLimit;
+  QueryExecutionContext* testQec = ad_utility::testing::getQec();
+  AllocatorWithLimit<Id> testAllocator = ad_utility::testing::makeAllocator();
+
+  // Create and run a HTTP server, which accepts queries of exactly the
+  // following form, as a POST request, with content-type header
+  // `application/sparql-query` and accept header `text/tab-separated-values`:
+  //
+  // SELECT <list of variables, separated by space> WHERE {
+  //   <result of query as TSV, without header row>
+  // }
+  //
+  // It then returns a TSV which consists of a header row (created from the
+  // SELECT variables), followed by the TSV from the query body. Understand why
+  // the test below contains a TSV result in the query body, instead of a proper
+  // SPARQL query body: We don't want to test whether our test server can
+  // evaluate a real SPARQL query, we want to test whether the SERVICE operation
+  // does the right thing with the TSV sent from a remote endpoint. Note that
+  // the `SELECT ... WHERE` part is added by the `Service` operation, hence we
+  // do want to test that part.
+  //
+  // If the query does not exactly have the form described above, the test
+  // fails.
+  TestHttpServer httpServer([](auto request,
+                               auto&& send) -> boost::asio::awaitable<void> {
+    EXPECT_EQ(request.method(), http::verb::post);
+    EXPECT_EQ(request.base()[http::field::content_type],
+              "application/sparql-query");
+    EXPECT_EQ(request.base()[http::field::accept], "text/tab-separated-values");
+
+    // Parse the query.
+    std::string_view testSparqlQuery = request.body();
+    static constexpr auto testQueryRegex = ctll::fixed_string(
+        "^\\s*SELECT\\s+(.*)\\s+WHERE\\s*\\{\\s*((.*)\\n)\\s*\\}$");
+    auto match = ctre::search<testQueryRegex>(testSparqlQuery);
+    EXPECT_TRUE(match);
+    // Why do I need `template` before `get<...>` in the following? I don't need
+    // it in other places in the code (e.g, in `util/GeoSparqlHelpers.cpp`).
+    std::string_view variables = match.template get<1>();
```

Comment:
If that is really as you say, then it is quite confusing, because it seems that you do the exact same thing.
Which compiler complains about what?
Otherwise just leave it.